### PR TITLE
Integrate hipsolver batched linalg drivers 

### PR DESCRIPTION
**Pull request by alugorey**
Enables the following tests for ROCm along with support for various batched linalg drivers:
test_inverse_errors_large_cuda*
test_qr_batched_cuda*
test_linalg_solve_triangular_large_cuda*
test_ormqr_cuda_complex*
test_householder_product_cuda_complex*
test_geqrf_cuda_complex*

### CODE REVIEW

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp
Comment (lezcano) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1842]: Should we just delete this if, as you did in the other changes below?

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp
Comment (lezcano) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1625]: Why not set this to true?

File: aten/src/ATen/cuda/CUDABlas.h
Comment (lezcano) [inline aten/src/ATen/cuda/CUDABlas.h:433]: or simply `elif defined(CUDART_VERSION)`

**Review by lezcano — COMMENTED** [review]
Left a number of comments and questions.

You should probably also activate the relevant OpInfo tests

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp
Comment (alugorey) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1842]: I believe this #ifdef is needed because in the event that we are not using CUDA or ROCM, then the logic just defaults to take the magma route on line 1874. At least that is my understanding of the code. Please correct me if I am wrong.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp
Comment (alugorey) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:407]: Done. added #elif to check for lower version and if so, set it to the previous value of `rocblas_operation_none`

**Comment by alugorey** [comment]
Hi @nikitaved , @lezcano , This is ready for final review. It is failing the following 3 unrelated and unstable test cases.
cudnn_extension error: trunk / linux-focal-rocm5.6-py3.8 / test (default, 2, 3, linux.rocm.gpu, unstable) (push) 
torch dynamo error: pull / linux-focal-py3.8-gcc7 / test (distributed, 2, 2, linux.2xlarge, unstable) (pull_request) 
torch dynamo error: pull / linux-bionic-cuda11.8-py3.10-gcc9 / test (distributed, 3, 3, linux.8xlarge.nvidia.gpu, unstable) (pull_request) Failing after 11m

File: aten/src/ATen/cuda/CUDABlas.h
Comment (lezcano) [inline aten/src/ATen/cuda/CUDABlas.h:315]: ditto. Grep for all these and write `elif define` as it's easier to follow

**Review by lezcano — COMMENTED** [review]
This is looking quite good, but I'd like to make sure we're not breaking some non-standard build.

There are many changes that are not semantics preserving. I pointed out a few.

Now, there are many cases where we do things like `if ROCM ... elif cusolver ... else`. Is it possible to build without ROCM or cusolver support? cc @malfet

If we always have either cusolver or hipsolver, then a fair amount of the code could be simplified. In particular, we would always have that `defined(CUDART_VERSION) || defined(ROCM_VERSION)` and quite a bit of the code could be simplified.

If there are cases where we don't have cusolver or hipsolver, then we should try to build in those cases with this patch, as we may be breaking them.

Once we know the answer to the question above, we should write a note for future developers somewhere.

File: aten/src/ATen/cuda/Exceptions.h
Comment (malfet) [inline aten/src/ATen/cuda/Exceptions.h:67]: Why this is needed? Shouldn't hipifier just replace `TORCH_CUDABLAS_CHECK` with TORCH_HIPBLAS_CHECK?

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (malfet) [inline aten/src/ATen/cuda/CUDABlas.cpp:999]: This looks like a verbatim copy of https://github.com/pytorch/pytorch/blob/854fe470cdc78873268bf3b8c8d392173f2aac8e/aten/src/ATen/cuda/CUDABlas.cpp#L985-L995

which makes me wonder, why hipifier can not take care of that one?

**Review by malfet — CHANGES_REQUESTED** [review]
Large chunks of code seems to copy-n-pasted in this PR instead of modifying hipifier to take care about them.
Please either elaborate in PR description, why HIPifier can not be used here, or better adjust it to take care of the duplications. Also, if you are adding/removing some constraints that are not specific to ROCM, please mention in PR description why they should no longer apply

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (alugorey) [reply aten/src/ATen/cuda/CUDABlas.cpp:1321]: I believe you are correct. There could be instances where neither solver is installed and it instead uses magma or some other LAPACK library. Did I add logic that assumes cusolver is installed when `USE_ROCM` is False?

File: aten/src/ATen/cuda/CUDABlas.h
Comment (alugorey) [reply aten/src/ATen/cuda/CUDABlas.h:315]: I'm slightly confused as to why we want `#elif defined(CUDART_VERSION)` as opposed to `#else`? My initial rationale behind else and not elif was because the existing CUDABLAS code was there by default and there was no macro conditional protecting it. If I add elif defined, we are now removing those pieces of code that have neither hipsolver or cusolver and could break non-standard builds. I think you mentioned this issue in a previous comment.

Comment (alugorey) [reply aten/src/ATen/cuda/CUDABlas.h:315]: woops ignore previous comment.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp
Comment (alugorey) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:2479]: OOps, yea this should definitely have been changed to USE_LINALG_SOLVER instead of being removed

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (lezcano) [reply aten/src/ATen/cuda/CUDABlas.cpp:1321]: Let's see if @malfet's approach is possible, which would heavily reduce the amount of code needed. If not, it'd be good to put together a build without rocm or cusolver (if such a build exists, I'm not sure) and then try to build and run the tests, see if it's correct.

cc @IvanYashchuk who surely knows the answer as to whether we can build without cusolver and hipsolver.

File: test/test_linalg.py
Comment (alugorey) [reply test/test_linalg.py:3628]: woops, that was leftover from internal testing

File: test/test_linalg.py
Comment (alugorey) [reply test/test_linalg.py:4060]: The comment is inaccurate. With the recent CI upgrade to rocm 5.6 and the code encompassed in this PR, there is no longer a memory access bug. However, while the test passes locally for us, it still is failing in CI due to a "Tensor-likes are not close" error where a single entry fails the comparison by about 0.002. The root cause could be the difference between actual systems being run, but we don't know for sure. It is on my list of TODOs to investigate further but for now I think it is accurate to remove the comment about a memory access bug and instead opting for the "doesn't work on rocm (yet) decorator".

**Comment by alugorey** [comment]
@malfet Thank you for your comments! I am in the mix of addressing them now and seeing if we can get by with hipification. I think the original motivation for not using hipification was because some hipblas and rocblas types were not easily interchangeable. A problem that I think may have been addressed in later rocm versions. I am looking into that now.

Also, any instance of this PR removing constraints that are NOT ROCM specific is an oversight on my part. I have addressed the locations you have pointed out in my current working version. Please continue to scrutinize every bit of this PR as I believe we all want this to be as high-quality as possible. Thanks again for your time in reviewing this :) new patch set will come in soon!

**Comment by alugorey** [comment]
@malfet @lezcano I have investigated the hipification request. Unfortunately it uncovered a known limitation. Currently, `cublas` types and api are hipified to `rocblas`. However, the api introduced here require `cublas` to be converted to `hipblas`. The issue is that there are many BLAS calls unrelated to the changes in this PR that still require `rocblas`. If I were to blindly hipify all instances of `cublas` to `hipblas` I would break a whole lot of code unrelated to this PR. This is a known issue and is currently being worked on, no hard ETA at the moment. The duplication of code in this PR that could seemingly have otherwise have been hipified is due to this limitation. I am aware of the pieces of work required to remove this limitation and will submit a new PR once that code gets merged. Thoughts?

**Comment by lezcano** [comment]
Given the size of this PR, it may be simpler to wait until those fixes land, and then rebase this one on top of that one and heavily simplify it.

**Comment by alugorey** [comment]
Understood. I'm going to push up the things I fixed in terms of the other comments in order to save my place.

**Comment by alugorey** [comment]
Here is the work in question: https://github.com/pytorch/pytorch/pull/105881

**Comment by alugorey** [comment]
The contents of this PR were brought in as part of https://github.com/pytorch/pytorch/pull/105881. The lingering functionality not encompasses in that PR will be brought in here: https://github.com/pytorch/pytorch/pull/106620
Closing this issue @malfet @lezcano

**Pull request closed without merging**
